shared/src/ledger/ibc/handler.rs

```diff
@@ -178,11 +231,601 @@ pub enum Error {
+        let channel_key = storage::channel_key(&port_channel_id);
+        let value = self.read_ibc_data(&channel_key).ok_or_else(|| {
+            Error::Channel(format!(
+                "The channel doesn't exist: Port/Channel {}",
+                port_channel_id
+            ))
+        })?;
+        let mut channel =
+            ChannelEnd::decode_vec(&value).map_err(Error::Decoding)?;
+        close_channel(&mut channel);
+        self.write_ibc_data(
+            &channel_key,
+            channel.encode_vec().expect("encoding shouldn't fail"),
+        );
+
+        let event = make_close_init_channel_event(msg);
+        self.emit_ibc_event(event);
+
+        Ok(())
+    }
+
+    /// Close the channel for ChannelCloseConfirm
+    fn close_confirm_channel(
+        &self,
+        msg: &MsgChannelCloseConfirm,
+    ) -> Result<()> {
+        let port_channel_id =
+            port_channel_id(msg.port_id.clone(), msg.channel_id.clone());
+        let channel_key = storage::channel_key(&port_channel_id);
+        let value = self.read_ibc_data(&channel_key).ok_or_else(|| {
+            Error::Channel(format!(
+                "The channel doesn't exist: Port/Channel {}",
+                port_channel_id
+            ))
+        })?;
+        let mut channel =
+            ChannelEnd::decode_vec(&value).map_err(Error::Decoding)?;
+        close_channel(&mut channel);
+        self.write_ibc_data(
+            &channel_key,
+            channel.encode_vec().expect("encoding shouldn't fail"),
+        );
+
+        let event = make_close_confirm_channel_event(msg);
+        self.emit_ibc_event(event);
+
+        Ok(())
+    }
+
+    /// Send a packet
+    fn send_packet(&self, data: &PacketSendData) -> Result<()> {
+        // get and increment the next sequence send
+        let port_channel_id = port_channel_id(
+            data.source_port.clone(),
+            data.source_channel.clone(),
+        );
+        let seq_key = storage::next_sequence_send_key(&port_channel_id);
+        let sequence = self.get_and_inc_sequence(&seq_key)?;
+
+        // store the commitment of the packet
+        let packet = data.packet(sequence);
+        let commitment_key = storage::commitment_key(
+            &data.source_port,
+            &data.source_channel,
+            packet.sequence,
+        );
+        let commitment = commitment(&packet);
+        self.write_ibc_data(&commitment_key, commitment.as_bytes());
+
+        let event = make_send_packet_event(packet);
+        self.emit_ibc_event(event);
+
+        Ok(())
+    }
+
+    /// Receive a packet
+    fn receive_packet(&self, msg: &MsgRecvPacket) -> Result<()> {
+        // store the receipt
+        let receipt_key = storage::receipt_key(
+            &msg.packet.destination_port,
+            &msg.packet.destination_channel,
+            msg.packet.sequence,
+        );
+        // write 1 as a receipt
+        self.write_ibc_data(&receipt_key, vec![1_u8]);
```

Comment:
this value is not significant, right? Could it be an empty array?

Reply:
Good point! Yes, the value isn't significant. In ibc-go, `1` is written as a receipt.
But it isn't clear. I will add `PacketReceipt` in #735.
https://github.com/cosmos/ibc-go/blob/147d9859f56f079edf96f4d7eea4c1261129bcf2/modules/core/04-channel/keeper/keeper.go#L173-L176